CI: Add workflow to do a simple smoke test from a release

mook-as opened the pull request:
This is used to verify that our artifacts in a release are correct (and we can ship them).  It checks that signatures look sane.

This currently just starts the application then quits.

jandubois commented on .github/workflows/smoke-test.yaml:
```diff
@@ -0,0 +1,182 @@
+          echo "Failed to find release tag" >&2
+          exit 1
+        fi
+        # For macOS, currently only x86_64 supports nested virtualization
```
Instead of skipping macOS on aarch64 completely, we could still do the download and unpack tests.

Maybe there is also a way to run the app in the "blocked" state (pretending there is a snapshot operation in progress), so the GUI will start, but not run the Lima VM. Not a full test, but better than nothing.

jandubois replied:
> Maybe there is also a way to run the app in the "blocked" state (pretending there is a snapshot operation in progress), so the GUI will start, but not run the Lima VM. Not a full test, but better than nothing.

On second thoughts this doesn't sound too useful. Just unpacking the bits and checking the signatures is enough.

jandubois reviewed (commented):
I don't have the patience to review the script itself right now, and I think we won't need it until we do the next release, so I will review other PRs instead right now.

The workflow seems fine to me, except it could probably share steps with the `bats` workflow (see #6692).

The other issue I have is that we should still be able to do some reduced testing on macOS for aarch64.

mook-as commented:
Added checking signature of macos-aarch64.

Also, converted the script to shell (instead of NodeJS); it doesn't seem any shorter, though.  Note that this will currently fail on Windows because the shell scripts runs under msys; this means that this PR includes a fix for #1807 but we don't have any signed releases with that change (so any existing signed releases will fail to start).

jandubois commented on .github/workflows/smoke-test.yaml:
```diff
@@ -0,0 +1,186 @@
+        # https://github.com/actions/runner-images/issues/9460
+        gh release download "$RELEASE_TAG" \
+          --repo ${{ github.repository }} \
+          --pattern '*.x86_64.dmg' \
```
Shouldn't this now also download the `aarch64` installers?

jandubois commented on .github/workflows/smoke-test.yaml:
```diff
@@ -0,0 +1,186 @@
+      fail-fast: false
+      matrix:
+        include:
+        - { platform: macos, runs-on: macos-13, pattern: '*.aarch64.dmg' }
```
This runs the smoke test on `aarch64` but not on `x86_64` anymore.

I also don't see any logic in the shell script that would just run the checksum/signature checks for `aarch64` but skips starting the application.

jandubois reviewed (commented):
> Added checking signature of macos-aarch64.

I don't understand how this is supposed to work. It looks like it does a full smoke test on aarch64 and doesn't test x86_64 anymore at all.

mook-as commented:
@jandubois Sorry, I missed pushing some local changes to actually make this work; updated.

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,223 @@
+
+set -o errexit -o nounset
+
+# All commands in the cleanups array will be run on exit.  They must be plain
+# strings that will be passed to eval
+cleanups=()
```
Is it for testing the script on a non-CI machine?

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,223 @@
+    mkdir -p "$destApp"
+    cleanups+=("rm -rf '$destApp'")
+
+    tar -cC "$srcApp" . | tar -xC "$destApp"
```
This looks more complicated than `cp -R …`. The only reason I can think of where `tar` is better is if the source can contain hard links, which should not be the case here. Are there other reasons?

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,223 @@
+    xattr -d -r -s -v com.apple.quarantine "$destApp"
+
+    if [[ "$(uname -m)" =~ arm ]]; then
+        # For macOS, currently only x86_64 supports nested virtualization
```
```suggestion
        # For macOS, currently only x86_64 runners supports nested virtualization
```

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,223 @@
+    deadline=$(( $(date +%s) + 10 * 60 ))
+
+    while [[ $(date +%s) -lt $deadline ]]; do
+        state=$(MSYS2_ARG_CONV_EXCL='*' "$RDCTL" api /v1/backend_state || echo '{"vmState": "UNREADY"}')
```
Is there any reason not to set this once on script start instead of prefixing individual commands with it?

```shell
export MSYS2_ARG_CONV_EXCL='*'
```

At first I thought this line was Win32-specific until I realized that the variable will just be ignored on other platforms.

jandubois reviewed (approved):
Thanks, LGTM

mook-as reviewed (commented):
Updated to address comments, and also I reversed the condition when waiting for msiexec to quit.

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,226 @@
+    deadline=$(( $(date +%s) + 10 * 60 ))
+    while [[ $(date +%s) -lt $deadline ]]; do
+        if tasklist.exe /FI "ImageName eq msiexec.exe" | grep msiexec; then
+            printf "Waiting for msiexec: %s/%s\n" "$(date)" "$(date --date="@$deadline")"
```
```suggestion
            printf "Waiting for msiexec finish: %s/%s\n" "$(date)" "$(date --date="@$deadline")"
```

mook-as replied:
Changed.

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,226 @@
+        else
+            break
+        fi
+    done
```
Shouldn't the script throw an error if the deadline has been reached and `msiexec` is still running? Instead of just trying to run the app?

mook-as replied:
Good catch; check added.

jandubois reviewed (approved):
Thanks, LGTM.

I just have one error message clarification, and maybe some error handling when `msiexec` exceeds the deadline.

jandubois commented on .github/workflows/smoke-test/smoke-test.sh:
```diff
@@ -0,0 +1,231 @@
+            break
+        fi
+    done
+    if [[ -z "$completed" ]]; then
```
Doesn't this fail because of `set -o nounset`?

```suggestion
    if [[ -z "${completed:-}" ]]; then
```

mook-as replied:
No, because it's declared on line 146.

jandubois replied:
It is declared, but not bound to a value. Declaring doesn't set it to the empty string:

```bash
#!/usr/bin/env bash

set -o errexit -o nounset

foo() {
    local completed
    if [[ -z "$completed" ]]; then
        echo not completed
    fi
    echo foo
}

foo
```

Fails:

```console
$ ./nounset.sh
./nounset.sh: line 7: completed: unbound variable
```

You would have to declare it like `local completed=""` to make it work if you don't want to change the condition to `[[ -z "${completed:-}" ]]`.

mook-as replied:
Ugh. I tested this before replying, but I specifically tested the version shipped with macOS… the newer version changes the behaviour.

This works on `3.2.57(1)-release`
But it fails on `5.2.15(1)-release`.

I'll go fix it.

mook-as replied:
Done.

jandubois reviewed (changes requested):
Looks good, except for the one unbound variable issue.

jandubois reviewed (approved):
Thanks, LGTM
